File: packages/prebid_mobile_sdk_max/android/src/main/kotlin/com/prebid/prebid_mobile_sdk_max/MaxInterstitialManager.kt
package com.prebid.prebid_mobile_sdk_max

import android.app.Activity
import com.applovin.mediation.MaxAd
import com.applovin.mediation.MaxAdListener
import com.applovin.mediation.MaxError
import com.applovin.mediation.adapters.prebid.utils.MaxMediationInterstitialUtils
import com.applovin.mediation.ads.MaxInterstitialAd
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.prebid.mobile.api.data.AdUnitFormat
import org.prebid.mobile.api.mediation.MediationInterstitialAdUnit
import java.util.EnumSet

/// Handles MAX-mediated interstitials over the
/// `prebid_mobile_sdk_max/interstitial` method channel. Each ad is keyed by an
/// `adId` allocated on the Dart side; native events are pushed back over the
/// same channel.
class MaxInterstitialManager(
    messenger: BinaryMessenger,
    private val activityProvider: () -> Activity?,
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, "prebid_mobile_sdk_max/interstitial")

    private class Holder(
        val adUnit: MediationInterstitialAdUnit,
        val interstitial: MaxInterstitialAd,
    )

    private val ads = mutableMapOf<Long, Holder>()

    init {
        channel.setMethodCallHandler(this)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        ads.values.forEach {
            it.adUnit.destroy()
            it.interstitial.destroy()
        }
        ads.clear()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val args = call.arguments as? Map<*, *>
        val adId = (args?.get("adId") as? Number)?.toLong()

        when (call.method) {
            "load" -> {
                val activity = activityProvider()
                if (activity == null) {
                    result.error("no_activity", "No attached Activity to load the interstitial", null)
                    return
                }
                if (adId == null) {
                    result.error("no_ad_id", "Missing adId", null)
                    return
                }
                val configId = args?.get("configId") as? String ?: ""
                val maxAdUnitId = args?.get("maxAdUnitId") as? String ?: ""
                val isVideo = args?.get("isVideo") as? Boolean ?: false

                val interstitial = MaxInterstitialAd(maxAdUnitId, activity)
                interstitial.setListener(object : MaxAdListener {
                    override fun onAdLoaded(ad: MaxAd) = send(adId, "onAdLoaded")
                    override fun onAdLoadFailed(adUnitId: String, error: MaxError) =
                        send(adId, "onAdFailed", error.message)
                    override fun onAdDisplayed(ad: MaxAd) = send(adId, "onAdDisplayed")
                    override fun onAdDisplayFailed(ad: MaxAd, error: MaxError) =
                        send(adId, "onAdFailed", error.message)
                    override fun onAdHidden(ad: MaxAd) = send(adId, "onAdClosed")
                    override fun onAdClicked(ad: MaxAd) = send(adId, "onAdClicked")
                })

                val mediationUtils = MaxMediationInterstitialUtils(interstitial)
                val format = if (isVideo) AdUnitFormat.VIDEO else AdUnitFormat.BANNER
                val adUnit = MediationInterstitialAdUnit(
                    activity,
                    configId,
                    EnumSet.of(format),
                    mediationUtils,
                )
                ads[adId] = Holder(adUnit, interstitial)

                adUnit.fetchDemand {
                    interstitial.loadAd()
                }
                result.success(null)
            }

            "show" -> {
                adId?.let { ads[it]?.interstitial }?.let { ad ->
                    if (ad.isReady) ad.showAd()
                }
                result.success(null)
            }

            "destroy" -> {
                adId?.let { ads.remove(it) }?.let {
                    it.adUnit.destroy()
                    it.interstitial.destroy()
                }
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun send(adId: Long, event: String, error: String? = null) {
        val payload = mutableMapOf<String, Any?>("adId" to adId)
        if (error != null) payload["error"] = error
        channel.invokeMethod(event, payload)
    }
}

File: packages/prebid_mobile_sdk_max/ios/prebid_mobile_sdk_max/Sources/prebid_mobile_sdk_max/MaxInterstitialManager.swift
import Flutter
import UIKit
import PrebidMobile
import PrebidMobileMAXAdapters
import AppLovinSDK

/// Handles MAX-mediated interstitials over the
/// `prebid_mobile_sdk_max/interstitial` method channel. Each ad is keyed by an
/// `adId` allocated on the Dart side; native events are pushed back over the
/// same channel.
class MaxInterstitialManager: NSObject, MAAdDelegate {

    private let channel: FlutterMethodChannel

    private var adUnits: [Int: MediationInterstitialAdUnit] = [:]
    private var mediationDelegates: [Int: MAXMediationInterstitialUtils] = [:]
    private var interstitials: [Int: MAInterstitialAd] = [:]

    init(messenger: FlutterBinaryMessenger) {
        channel = FlutterMethodChannel(
            name: "prebid_mobile_sdk_max/interstitial",
            binaryMessenger: messenger
        )
        super.init()
        channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call, result)
        }
    }

    private func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let adId = args?["adId"] as? Int

        switch call.method {
        case "load":
            guard let adId = adId else {
                result(FlutterError(code: "no_ad_id", message: "Missing adId", details: nil))
                return
            }
            let configId = args?["configId"] as? String ?? ""
            let maxAdUnitId = args?["maxAdUnitId"] as? String ?? ""
            let isVideo = args?["isVideo"] as? Bool ?? false

            // 1. Create the MAX interstitial + Prebid mediation utils + ad unit.
            let interstitial = MAInterstitialAd(adUnitIdentifier: maxAdUnitId)
            let mediationDelegate = MAXMediationInterstitialUtils(interstitialAd: interstitial)
            let adUnit = MediationInterstitialAdUnit(
                configId: configId,
                mediationDelegate: mediationDelegate
            )
            adUnit.adFormats = isVideo ? [.video] : [.banner]

            interstitial.delegate = self
            adUnits[adId] = adUnit
            mediationDelegates[adId] = mediationDelegate
            interstitials[adId] = interstitial

            // 2. Fetch demand, then load the MAX interstitial.
            adUnit.fetchDemand { [weak self] _ in
                self?.interstitials[adId]?.load()
            }
            result(nil)

        case "show":
            if let adId = adId, let interstitial = interstitials[adId], interstitial.isReady {
                interstitial.show()
            }
            result(nil)

        case "destroy":
            if let adId = adId {
                interstitials.removeValue(forKey: adId)
                adUnits.removeValue(forKey: adId)
                mediationDelegates.removeValue(forKey: adId)
            }
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func send(_ adId: Int, _ event: String, error: String? = nil) {
        var payload: [String: Any] = ["adId": adId]
        if let error = error {
            payload["error"] = error
        }
        channel.invokeMethod(event, arguments: payload)
    }

    // MARK: - MAAdDelegate

    func didLoad(_ ad: MAAd) {
        // MAAd does not carry the ad unit instance; resolve by unit identifier.
        if let adId = adId(forAdUnitIdentifier: ad.adUnitIdentifier) {
            send(adId, "onAdLoaded")
        }
    }

    func didFailToLoadAd(forAdUnitIdentifier adUnitIdentifier: String, withError error: MAError) {
        if let adId = adId(forAdUnitIdentifier: adUnitIdentifier) {
            send(adId, "onAdFailed", error: error.message)
        }
    }

    func didDisplay(_ ad: MAAd) {
        if let adId = adId(forAdUnitIdentifier: ad.adUnitIdentifier) {
            send(adId, "onAdDisplayed")
        }
    }

    func didHide(_ ad: MAAd) {
        if let adId = adId(forAdUnitIdentifier: ad.adUnitIdentifier) {
            send(adId, "onAdClosed")
        }
    }

    func didClick(_ ad: MAAd) {
        if let adId = adId(forAdUnitIdentifier: ad.adUnitIdentifier) {
            send(adId, "onAdClicked")
        }
    }

    func didFail(toDisplay ad: MAAd, withError error: MAError) {
        if let adId = adId(forAdUnitIdentifier: ad.adUnitIdentifier) {
            send(adId, "onAdFailed", error: error.message)
        }
    }

    /// Resolves the owning `adId` from a MAX ad unit identifier.
    private func adId(forAdUnitIdentifier identifier: String) -> Int? {
        for (adId, interstitial) in interstitials
        where interstitial.adUnitIdentifier == identifier {
            return adId
        }
        return nil
    }
}
